vtex/cli-plugin-functions
# Implements a proxy, serving content from gatsby and functions

#### What is the purpose of this pull request?
It adds a proxy server to serve the Gatsby store and functions from the same server, using the same port

#### What problem is this solving?

We need to support serverless functions on StoreFramework JAMStack. This PR allows developers to test their functions locally in a similar environment we will provide on production.

#### How should this be manually tested?
You will need a Gatsby store with a directory `api` containing functions. You can use this project as an example: [vtex-site/storefaas.store](https://github.com/vtex-sites/storefaas.store).

1. Clone this repo ([cli-plugin-functions](https://github.com/vtex/cli-plugin-functions)) and run `yarn link`.
1. Clone [`vtex/toolbelt`](https://github.com/vtex/toolbelt), checkout the `add/serverless-functions` branch.
1. Now run `yarn link @vtex/cli-plugin-functions` on the `vtex/toolbelt` project, and install other dependencies running `yarn`.
1. Run `yarn watch` on the `vtex/toolbelt`. Please note you may need to add `$HOME/.vtex/dev/bin` to your PATH, [check this document](https://github.com/vtex/toolbelt/blob/master/docs/contributing.md).
1. Test the command on a Gatsby store with `vtex-test functions`

#### Screenshots or example usage
Run `vtex-test functions` on a Gatsby store with functions.

It will create three servers:
- http://localhost:3000, the Express functions server
- http://localhost:8000, the Gatsby server
- http://localhost:8080, the proxy server

The proxy server will rewrite `^/functions/` to `/` on the functions server. Any other path will have its request forwarded to the Gatsby server.

Open the proxy endpoint (http://localhost:8080) on a browser and click on the banner. It should show an alert with the response from a serverless function. 

![Test the serverless function by clicking on the banner](https://user-images.githubusercontent.com/8618619/105712896-3bea2380-5ef9-11eb-815c-f3ebbc3ca31a.png)

#### Types of changes

#### Chores checklist
